packages/react/src/components/SelectUsersModal/SelectUsersModal.jsx

```diff
@@ -0,0 +1,266 @@
+      isCategory: depth === 0 && !isEmpty(children),
+      disabled: renderDisabledState({ user, parents, selected: selectedUsers, depth }),
+      children,
+    };
+  });
+
+const SelectUsersModal = ({ open, onClose, onSubmit, users, initialSelectedUsers, i18n }) => {
+  const mergedI18n = { ...defaultProps.i18n, ...i18n };
+
+  const [selectedUsers, setSelectedUsers] = useState(initialSelectedUsers);
+
+  const handleRemove = (row) => {
+    setSelectedUsers(selectedUsers.filter((s) => s !== row));
+  };
+
+  const handleAdd = (row) => {
+    setSelectedUsers([row].concat(...selectedUsers));
+  };
+
+  const displayAllUsersList = (unmodifiedUsers, selection = []) => {
+    return mapUsers(unmodifiedUsers, {
+      selectedUsers: selection,
+      renderDisabledState: ({ user, selected, parents }) => {
+        const isInSelected = selected.filter((s) => itemsAreEqual(s, user)).length > 0;
+        const isChildOfSelected =
+          selected.filter((u) => parents.filter((p) => itemsAreEqual(p, u)).length > 0).length > 0;
+
+        return isInSelected || isChildOfSelected;
+      },
+      renderRowActions: ({ user, selected, parents, depth }) => {
+        const { username } = user;
+        const isInSelected = selected.filter((s) => itemsAreEqual(s, user)).length > 0;
+        const isChildOfSelected =
+          selected.filter((u) => parents.filter((p) => itemsAreEqual(p, u)).length > 0).length > 0;
+
+        // Don't render actions on the top level unless it's a user. TODO: is this supposed to support a flat list, too?
+        const isCategory = depth === 0 && user && !user.username;
+
+        if (isInSelected || isChildOfSelected || isCategory) {
+          return null;
+        }
+
+        return () => [
+          <Button
+            key={`${username}-list-item-button-${depth}`}
+            style={{ color: 'black' }}
+            role="button"
+            aria-label="Add"
+            renderIcon={ArrowRight16}
+            hasIconOnly
+            kind="ghost"
+            size="small"
+            onClick={() => handleAdd(user)}
+            iconDescription={mergedI18n.addUser}
+          />,
+        ];
+      },
+    });
+  };
+
+  const displaySelectedUsersList = (unmodifiedUsers) => {
+    return mapUsers(unmodifiedUsers, {
+      renderDisabledState: () => false,
+      renderRowActions: ({ user, depth }) => {
+        const { username } = user;
+
+        if (depth === 0) {
+          return () => [
+            <Button
+              key={`${username}-list-item-button-${depth}`}
+              style={{ color: 'black' }}
+              aria-label="Remove"
+              renderIcon={ArrowLeft16}
+              hasIconOnly
+              kind="ghost"
+              size="small"
+              onClick={() => handleRemove(user)}
+              iconDescription={mergedI18n.removeUser}
+            />,
+          ];
+        }
+
+        return null;
+      },
+    });
+  };
+
+  const flattenUsers = (results, user) => {
```

Comment:
I'm not sure how many users you're thinking of managing but you might need to memoize and useCallback if you're rendering a bunch